eos/effects/adaptivearmorhardener.py

```diff
@@ -2,17 +2,207 @@
-            damagePatternModifier = getattr(damagePattern,attrDamagePattern)/float(sum((damagePattern.emAmount,damagePattern.thermalAmount,damagePattern.kineticAmount,damagePattern.explosiveAmount)))
-            modifiedResistModifier = (1-(((1-module.getModifiedItemAttr(attr))*4)*(damagePatternModifier)))
-            module.forceItemAttr(attr, modifiedResistModifier)
-
-        fit.ship.multiplyItemAttr(attr, module.getModifiedItemAttr(attr), stackingPenalties=True, penaltyGroup="preMul")
+    damagePattern = fit.damagePattern
+
+    # Skip if there is no damage pattern. Example: projected ships or fleet boosters
+    if damagePattern:
+
+        # Populate a tuple with the damage profile modified by current armor resists.
+        # Build this up front as there are a number of values we only need to touch once
+        damagePattern_tuple = []
+
+        '''
+        # Bit of documentation for how the tuple is layed out
+        damagePattern_tuple.append(['RESIST TYPE',
+                SHIP ARMOR RESIST AMOUNT,
+                DAMAGE PATTERN AMOUNT,
+                MODIFIED DAMAGE PATTERN AMOUNT (SHIP ARMOR RESIST AMOUNT * DAMAGE PATTERN AMOUND),
+                ADAPTIVE RESIST AMOUNT,
+                ADAPTIVE RESIST AMOUNT (ORIGINAL))
+        '''
+
+        damagePattern_tuple.append(['Em',
+                fit.ship.getModifiedItemAttr('armorEmDamageResonance'),
+                damagePattern.emAmount,
+                damagePattern.emAmount*fit.ship.getModifiedItemAttr('armorEmDamageResonance'),
+                module.getModifiedItemAttr('armorEmDamageResonance'),
+                module.getModifiedItemAttr('armorEmDamageResonance')])
+
+        damagePattern_tuple.append(['Thermal',
+                fit.ship.getModifiedItemAttr('armorThermalDamageResonance'),
+                damagePattern.thermalAmount,
+                damagePattern.thermalAmount * fit.ship.getModifiedItemAttr('armorThermalDamageResonance'),
+                module.getModifiedItemAttr('armorThermalDamageResonance'),
+                module.getModifiedItemAttr('armorThermalDamageResonance')])
+
+        damagePattern_tuple.append(['Kinetic',
+                fit.ship.getModifiedItemAttr('armorKineticDamageResonance'),
+                damagePattern.kineticAmount,
+                damagePattern.kineticAmount*fit.ship.getModifiedItemAttr('armorKineticDamageResonance'),
+                module.getModifiedItemAttr('armorKineticDamageResonance'),
+                module.getModifiedItemAttr('armorKineticDamageResonance')])
+
+        damagePattern_tuple.append(['Explosive',
+                fit.ship.getModifiedItemAttr('armorExplosiveDamageResonance'),
+                damagePattern.explosiveAmount,
+                damagePattern.explosiveAmount*fit.ship.getModifiedItemAttr('armorExplosiveDamageResonance'),
+                module.getModifiedItemAttr('armorExplosiveDamageResonance'),
+                module.getModifiedItemAttr('armorExplosiveDamageResonance')])
```

Comment:
>That's where you're wrong. It will properly sort types that have different amounts of damage, but it won't touch the order of types that have the same amounts of damage. That doesn't matter for the amount of damage taken, thus why it's not a big deal, but it can cause the RAH numbers to appear significantly different than what happens in game. I prefer to be accurate if possible, and changing the order you add the data to the tuple is a simple way to make it more accurate.

No, it sorts it correctly.

Here's the Gnosis with only a RAH, damage pattern is `100 | 101 | 101 | 101`

Let's dump the resists in order (left being 0, right being 3):
```
2016-09-15 22:15:31,423 eos.effects.adaptivearmorhardener DEBUG    Presort (low -> high): Em | Thermal | Kinetic | Explosive
```
You can see the first time we dump resist names, the order matches how we populate the tuple.

We then run through our logic (including the sort):
2016-09-15 22:15:31,424 eos.effects.adaptivearmorhardener DEBUG    Adaptive Resists, Ship Resists, Modified Damage (EM|The|Kin|Exp) : 0.910000 | 0.790000 | 0.790000 | 0.910000 || 0.614250 | 0.533250 | 0.533250 | 0.614250 || 67.500000 | 68.175000 | 68.175000 | 68.175000
```

Now next loop lets dump the resist names again to see how it sorted them:
```
2016-09-15 22:15:31,424 eos.effects.adaptivearmorhardener DEBUG    Presort (low -> high): Em | Explosive | Kinetic | Thermal
```

Hey, look at that.  EM is lowest (as it should be, since it took the least damage).  The other three take equal modified damage, and it sorted them alphabetically as expected.

So.  All in all working properly.